### PR TITLE
fix(value): improve value setting workflow #58

### PR DESCRIPTION
**Pull request by jason-capsule42**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Fixes:** #58, #56, #42, #60

## Summary:

Significant changes to the workflow for setting the value. This corrects issues where the logic was previously recursive and causing the wrong values to end up in the input. This also corrects the behavior where a required datepicker was completely broken and no value could be set. 

This also adds correct validation when setting the value of the datepicker.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: demo/apiExamples.md [hunk 1]
```diff
@@ -8,10 +8,10 @@
 |---------------|---------------|-----------|-------------|--------------------------------------------------|
 | [centralDate](#centralDate) | `centralDate` | `Object`  |             | The date that determines the currently visible month. |
 | [disabled](#disabled)    | `disabled`    | `Boolean` |             | If set, disables the datepicker.                 |
-| [error](#error)       | `error`       | `Boolean` | false       | Sets a persistent error state (e.g. an error state returned from the server). |
 | [maxDate](#maxDate)     | `maxDate`     | `Date`    | "undefined" | Maximum date. All dates after will be disabled.  |
 | [minDate](#minDate)     | `minDate`     | `Date`    | "undefined" | Minimum date. All dates before will be disabled. |
 | [required](#required)    | `required`    | `Boolean` |             | Populates the `required` attribute on the input. Used for client-side validation. |
+| [validity](#validity)    | `validity`    | `String`  | "undefined" | Specifies the `validityState` this element is in. |
```
Comment (blackfalcon) [inline demo/apiExamples.md:14]: I am having an issue with using `validity` as an enumerable attribute. If the goal is to more closely bind ourselves to the HTML spec, this really should be `valitidystate`. 

Also, we should clearly state in the documentation that we are expecting string values that are consistent with the client-side validation expectations listed here.https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#client-side_validation

While I can see using this to our benefit internally, what is the benefit to the end user to have to pass in a specific state? 

I really like how Shoelace simplified this and handled this with an `invalid`  boolean. https://shoelace.style/components/input#:~:text=false-,invalid,-This%20will%20be

File: demo/apiExamples.md [hunk 2]
```diff
@@ -8,10 +8,10 @@
 |---------------|---------------|-----------|-------------|--------------------------------------------------|
 | [centralDate](#centralDate) | `centralDate` | `Object`  |             | The date that determines the currently visible month. |
 | [disabled](#disabled)    | `disabled`    | `Boolean` |             | If set, disables the datepicker.                 |
-| [error](#error)       | `error`       | `Boolean` | false       | Sets a persistent error state (e.g. an error state returned from the server). |
```
Comment (blackfalcon) [inline demo/apiExamples.md:11]: Ok, we are removing this, but we should say something about the fact that `validity` is talking to `input` and then addressing the error state. 

I understand that at this point the user does not have a way to inject a custom error message, so when that comes about, how do you see supporting that?

File: src/auro-datepicker.js [hunk 3]
```diff
@@ -182,49 +176,54 @@ class AuroDatePicker extends LitElement {
...
+    const validDateStrLength = 10;
+    if (this.value !== undefined) {
+      if ((!this.input.value || this.input.value.length === 0) && this.required) {
+        this.validity = 'valueMissing';
+        this.input.setCustomValidity = 'Please enter a date.';
+      } else {
+        const date = new Date(this.input.value);
+
+        if (this.input.value.length > 0 && !this.validDate(date)) {
+          this.validity = 'badInput';
```
Comment (blackfalcon) [inline src/auro-datepicker.js:202]: Is the idea here that the user would be injecting the validity state or this is all internal? When I am using the demo page, I am unable to get any of these error statements to appear.

File: demo/apiExamples.md (hunk 2 above)
Comment (blackfalcon) [reply demo/apiExamples.md:11]: The example for this is still in the docs.

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: src/auro-datepicker.js (hunk 3 above)
Comment (jason-capsule42) [reply src/auro-datepicker.js]: Datepicker (more accurately calendar) has current restrictions that prevent you from selecting a date before the minDate or after the maxDate. So it is impossible to use the datepicker normally and make those errors appear. If that restriction were removed from Calendar the errors can be made to show. 

Today, the only errors you can make happen are for the required flag and having no value (after first interaction) and tooShort (start typing a date but don't finish and then blur the input)

**Review by blackfalcon — CHANGES_REQUESTED** [review]
```
test/auro-datepicker.test.js:

 ❌ auro-datepicker > handles the required state being set
      AssertionError: expected null to equal 'valueMissing'
        at o.<anonymous> (test/auro-datepicker.test.js:59:53)

 ❌ auro-datepicker > respects maxDate setting
      AssertionError: expected null to equal 'valid'
        at o.<anonymous> (test/auro-datepicker.test.js:86:53)

 ❌ auro-datepicker > respects minDate setting
      AssertionError: expected null to equal 'valid'
        at o.<anonymous> (test/auro-datepicker.test.js:104:53)

Chrome: |██████████████████████████████| 1/1 test files | 4 passed, 3 failed

Code coverage: 88.78 %
Coverage for branches failed with 69.04 % compared to configured 80 %
View full coverage report at coverage/lcov-report/index.html

Finished running tests in 4.3s, failed to meet coverage threshold.

ERROR: "test" exited with 1.
```

**Comment by jason-capsule42** [comment]
> ```
> test/auro-datepicker.test.js:
> 
>  ❌ auro-datepicker > handles the required state being set
>       AssertionError: expected null to equal 'valueMissing'
>         at o.<anonymous> (test/auro-datepicker.test.js:59:53)
> 
>  ❌ auro-datepicker > respects maxDate setting
>       AssertionError: expected null to equal 'valid'
>         at o.<anonymous> (test/auro-datepicker.test.js:86:53)
> 
>  ❌ auro-datepicker > respects minDate setting
>       AssertionError: expected null to equal 'valid'
>         at o.<anonymous> (test/auro-datepicker.test.js:104:53)
> 
> Chrome: |██████████████████████████████| 1/1 test files | 4 passed, 3 failed
> 
> Code coverage: 88.78 %
> Coverage for branches failed with 69.04 % compared to configured 80 %
> View full coverage report at coverage/lcov-report/index.html
> 
> Finished running tests in 4.3s, failed to meet coverage threshold.
> 
> ERROR: "test" exited with 1.
> ```

None of these tests fail when properly linking the head of the dependent [auro-input PR](https://github.com/AlaskaAirlines/auro-input/pull/180).

Please check your work and try again.

**Review by blackfalcon — APPROVED** [review]

**Pull request merged**